src/pages/dataElements/form/useIsFieldValueUnique.ts

```diff
@@ -0,0 +1,55 @@
+import { useDataQuery } from '@dhis2/app-runtime'
+import i18n from '@dhis2/d2-i18n'
+import { useMemo } from 'react'
+import { Pager } from '../../../types/generated'
+
+const HAS_FIELD_VALUE_QUERY = {
+    dataElements: {
+        resource: 'dataElements',
+        params: (variables: Record<string, string>) => ({
+            pageSize: 1,
+            fields: 'id',
+            filter: [`${variables.field}:eq:${variables.value}`],
+        }),
+    },
+}
+
+interface QueryResponse {
+    dataElements: {
+        pager: Pager
+    }
+}
+
+export function useIsFieldValueUnique(field: string) {
+    const queryResult = useDataQuery<QueryResponse>(HAS_FIELD_VALUE_QUERY, {
+        lazy: true,
+        variables: { field, value: '' },
+    })
+
+    return useMemo(
+        () => ({
+            ...queryResult,
+            refetch: (value: string) => {
+                if (!value) {
+                    return
+                }
+
+                return queryResult.refetch({ field, value }).then(
+                    // We don't have access to app-runtime's internal type `JsonMap`,
+                    // so we have to ignore the type error
+                    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+                    // @ts-ignore
+                    (data: QueryResponse) => {
+                        if (data.dataElements.pager.total) {
+                            return i18n.t(
+                                'This {{field}} already exists, please choose antoher one',
+                                { field }
+                            )
+                        }
+                    }
+                )
+            },
+        }),
+        [field, queryResult]
+    )
+}
```

Comment:
I think this would be simpler if we just used `engine.query`, and return `useCallback`. RFF is keeping track of the `validating`-state (in meta.validating) based on the promise.

Reply:
Something like this:


```ts
export function useIsFieldValueUnique(field: string) {
    const engine = useDataEngine()

    const validate = useMemo(
        () =>
            memoize(async (value: string) => {
                if (!value) {
                    return undefined
                }
                const data = (await engine.query(HAS_FIELD_VALUE_QUERY, {
                    variables: { field, value },
                })) as unknown as QueryResponse
                if (data.dataElements.pager.total > 0) {
                    return i18n.t(
                        'This field requires a unique value, please choose another one'
                    )
                }
            }),
        [field, engine]
    )
    return useDebouncedCallback(validate, 200)
}

This will fix that the validation is run on every field-change - because we're memoizing the function. Also debouncing the callback to limit the network requests when typing.

Note that this means you need to wrap `InputFieldRFF` and provide `loading` to `InputFieldFF` based on `meta.validating`. (Or change InputFieldFF implementation). 
```

Reply:
I copied your suggestion, but I also did two more things:

* Every field now has `validateFields={[]}` because the loading spinner was still shown when changing another value
* Validation is triggered only on blur because validation on change (even when debounced and memoized) caused performance issues for me.